File: scripts/init_adapter.py
from __future__ import annotations

import argparse
import importlib.util
import sys
from pathlib import Path
from typing import Any

SCRIPT_DIR = Path(__file__).resolve().parent


def _load_resolver() -> Any:
    module_path = SCRIPT_DIR / "resolve_adapter.py"
    spec = importlib.util.spec_from_file_location("cautilus_resolver", module_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Could not load resolver from {module_path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


resolve_adapter = _load_resolver()


def _yaml_scalar(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    text = str(value)
    if text == "":
        return '""'
    if any(char in text for char in [":", "#", "{", "}", "[", "]", "\n"]) or text.strip() != text:
        escaped = text.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
        return f'"{escaped}"'
    return text


def _dump_yaml(value: Any, indent: int = 0) -> list[str]:
    prefix = " " * indent
    if isinstance(value, dict):
        lines: list[str] = []
        for key, item in value.items():
            if isinstance(item, (dict, list)):
                if isinstance(item, list) and not item:
                    lines.append(f"{prefix}{key}: []")
                    continue
                lines.append(f"{prefix}{key}:")
                lines.extend(_dump_yaml(item, indent + 2))
                continue
            lines.append(f"{prefix}{key}: {_yaml_scalar(item)}")
        return lines
    if isinstance(value, list):
        lines = []
        for item in value:
            if isinstance(item, dict):
                lines.extend(_dump_yaml_list_item(item, indent))
                continue
            if isinstance(item, list):
                lines.append(f"{prefix}-")
                lines.extend(_dump_yaml(item, indent + 2))
                continue
            lines.append(f"{prefix}- {_yaml_scalar(item)}")
        return lines
    return [f"{prefix}{_yaml_scalar(value)}"]


def _dump_yaml_list_item(value: dict[str, Any], indent: int) -> list[str]:
    prefix = " " * indent
    lines: list[str] = []
    first = True
    for key, item in value.items():
        if isinstance(item, (dict, list)):
            if isinstance(item, list) and not item:
                if first:
                    lines.append(f"{prefix}- {key}: []")
                else:
                    lines.append(f"{prefix}  {key}: []")
                first = False
                continue
            if first:
                lines.append(f"{prefix}- {key}:")
            else:
                lines.append(f"{prefix}  {key}:")
            lines.extend(_dump_yaml(item, indent + 4))
            first = False
            continue
        if first:
            lines.append(f"{prefix}- {key}: {_yaml_scalar(item)}")
            first = False
            continue
        lines.append(f"{prefix}  {key}: {_yaml_scalar(item)}")
    if not lines:
        lines.append(f"{prefix}- {{}}")
    return lines


def dump_yaml_document(data: dict[str, Any]) -> str:
    return "\n".join(_dump_yaml(data)) + "\n"


def scaffold_adapter(repo_root: Path, repo_name: str) -> dict[str, Any]:
    inferred = resolve_adapter.infer_repo_defaults(repo_root)
    return {
        "version": 1,
        "repo": repo_name,
        "evaluation_surfaces": [
            "prompt behavior",
            "workflow behavior",
        ],
        "baseline_options": [
            "baseline git ref in the same repo via {baseline_ref}",
        ],
        "required_prerequisites": [
            "choose a real baseline before comparing results",
        ],
        "preflight_commands": inferred.get("preflight_commands", []),
        "iterate_command_templates": inferred.get("iterate_command_templates", []),
        "held_out_command_templates": inferred.get("held_out_command_templates", []),
        "comparison_command_templates": inferred.get("comparison_command_templates", []),
        "full_gate_command_templates": inferred.get("full_gate_command_templates", []),
        "executor_variants": [],
        "artifact_paths": [],
        "report_paths": inferred.get("report_paths", []),
        "comparison_questions": [
            "Which scenarios improved, regressed, or stayed noisy after repeated samples?",
        ],
        "human_review_prompts": [
            {
                "id": "real-user",
                "prompt": "Where would a real user still judge the candidate worse despite benchmark wins?",
            }
        ],
        "iterate_samples_default": inferred.get("iterate_samples_default", 2),
        "held_out_samples_default": inferred.get("held_out_samples_default", 2),
        "comparison_samples_default": inferred.get("comparison_samples_default", 2),
        "full_gate_samples_default": inferred.get("full_gate_samples_default", 2),
        "history_file_hint": inferred.get("history_file_hint", "/tmp/workbench-history.json"),
        "profile_default": inferred.get("profile_default", "default"),
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--repo-root", type=Path, required=True)
    parser.add_argument("--repo-name")
    parser.add_argument("--adapter-name")
    parser.add_argument("--output", type=Path)
    parser.add_argument("--force", action="store_true")
    args = parser.parse_args()

    repo_root = args.repo_root.resolve()
    default_output = (
        Path(".agents/workbench-adapters") / f"{args.adapter_name}.yaml"
        if args.adapter_name
        else Path(".agents/workbench-adapter.yaml")
    )
    chosen_output = args.output or default_output
    output = chosen_output if chosen_output.is_absolute() else repo_root / chosen_output
    if output.exists() and not args.force:
        raise SystemExit(f"Adapter already exists at {output}. Use --force to overwrite.")

    output.parent.mkdir(parents=True, exist_ok=True)
    adapter = scaffold_adapter(repo_root, args.repo_name or repo_root.name)
    output.write_text(dump_yaml_document(adapter), encoding="utf-8")
    sys.stdout.write(f"{output}\n")


if __name__ == "__main__":
    main()

File: scripts/init_adapter.mjs
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import process from "node:process";
import { pathToFileURL } from "node:url";

import { inferRepoDefaults } from "./resolve_adapter.mjs";

function yamlScalar(value) {
	if (value === null) {
		return "null";
	}
	if (typeof value === "boolean") {
		return value ? "true" : "false";
	}
	if (typeof value === "number") {
		return String(value);
	}
	const text = String(value);
	if (text === "") {
		return "\"\"";
	}
	if (
		[":", "#", "{", "}", "[", "]", "\n"].some((char) => text.includes(char)) ||
		text.trim() !== text
	) {
		const escaped = text
			.replaceAll("\\", "\\\\")
			.replaceAll("\"", "\\\"")
			.replaceAll("\n", "\\n");
		return `"${escaped}"`;
	}
	return text;
}

function isPlainObject(value) {
	return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function appendYamlProperty(lines, prefix, key, item, nestedIndent) {
	if (Array.isArray(item) && item.length === 0) {
		lines.push(`${prefix}${key}: []`);
		return;
	}
	if (isPlainObject(item) || Array.isArray(item)) {
		lines.push(`${prefix}${key}:`);
		lines.push(...dumpYaml(item, nestedIndent));
		return;
	}
	lines.push(`${prefix}${key}: ${yamlScalar(item)}`);
}

function dumpYamlListItem(value, indent) {
	const prefix = " ".repeat(indent);
	const lines = [];
	let first = true;
	for (const [key, item] of Object.entries(value)) {
		const keyPrefix = first ? `${prefix}- ` : `${prefix}  `;
		appendYamlProperty(lines, keyPrefix, key, item, indent + 4);
		first = false;
	}
	if (lines.length === 0) {
		lines.push(`${prefix}- {}`);
	}
	return lines;
}

function dumpYamlArray(value, indent, prefix) {
	const lines = [];
	for (const item of value) {
		if (isPlainObject(item)) {
			lines.push(...dumpYamlListItem(item, indent));
			continue;
		}
		if (Array.isArray(item)) {
			lines.push(`${prefix}-`);
			lines.push(...dumpYaml(item, indent + 2));
			continue;
		}
		lines.push(`${prefix}- ${yamlScalar(item)}`);
	}
	return lines;
}

function dumpYaml(value, indent = 0) {
	const prefix = " ".repeat(indent);
	if (isPlainObject(value)) {
		const lines = [];
		for (const [key, item] of Object.entries(value)) {
			appendYamlProperty(lines, prefix, key, item, indent + 2);
		}
		return lines;
	}
	if (Array.isArray(value)) {
		return dumpYamlArray(value, indent, prefix);
	}
	return [`${prefix}${yamlScalar(value)}`];
}

export function dumpYamlDocument(data) {
	return `${dumpYaml(data).join("\n")}\n`;
}

function numericDefaults(inferred) {
	return {
		iterate_samples_default: inferred.iterate_samples_default ?? 2,
		held_out_samples_default: inferred.held_out_samples_default ?? 2,
		comparison_samples_default: inferred.comparison_samples_default ?? 2,
		full_gate_samples_default: inferred.full_gate_samples_default ?? 2,
	};
}

export function scaffoldAdapter(repoRoot, repoName) {
	const inferred = inferRepoDefaults(repoRoot);
	return {
		version: 1,
		repo: repoName,
		evaluation_surfaces: ["prompt behavior", "workflow behavior"],
		baseline_options: ["baseline git ref in the same repo via {baseline_ref}"],
		required_prerequisites: ["choose a real baseline before comparing results"],
		preflight_commands: inferred.preflight_commands ?? [],
		iterate_command_templates: inferred.iterate_command_templates ?? [],
		held_out_command_templates: inferred.held_out_command_templates ?? [],
		comparison_command_templates: inferred.comparison_command_templates ?? [],
		full_gate_command_templates: inferred.full_gate_command_templates ?? [],
		executor_variants: [],
		artifact_paths: [],
		report_paths: inferred.report_paths ?? [],
		comparison_questions: [
			"Which scenarios improved, regressed, or stayed noisy after repeated samples?",
		],
		human_review_prompts: [
			{
				id: "real-user",
				prompt: "Where would a real user still judge the candidate worse despite benchmark wins?",
			},
		],
		...numericDefaults(inferred),
		history_file_hint: inferred.history_file_hint ?? "/tmp/workbench-history.json",
		profile_default: inferred.profile_default ?? "default",
	};
}

function fail(message) {
	process.stderr.write(`${message}\n`);
	process.exit(1);
}

function usage(exitCode = 0) {
	const text = [
		"Usage:",
		"  node ./scripts/init_adapter.mjs --repo-root <dir> [--repo-name <name>] [--adapter-name <name>] [--output <path>] [--force]",
	].join("\n");
	const out = exitCode === 0 ? process.stdout : process.stderr;
	out.write(`${text}\n`);
	process.exit(exitCode);
}

function parseArgs(argv) {
	const options = {
		repoRoot: null,
		repoName: null,
		adapterName: null,
		output: null,
		force: false,
	};
	for (let index = 0; index < argv.length; index += 1) {
		const arg = argv[index];
		if (arg === "-h" || arg === "--help") {
			usage(0);
		}
		if (arg === "--force") {
			options.force = true;
			continue;
		}
		const field = {
			"--repo-root": "repoRoot",
			"--repo-name": "repoName",
			"--adapter-name": "adapterName",
			"--output": "output",
		}[arg];
		if (!field) {
			fail(`Unknown argument: ${arg}`);
		}
		const value = argv[index + 1];
		if (!value) {
			fail(`Missing value for ${arg}`);
		}
		options[field] = value;
		index += 1;
	}
	if (!options.repoRoot) {
		fail("--repo-root is required");
	}
	return options;
}

export function main(argv = process.argv.slice(2)) {
	const options = parseArgs(argv);
	const repoRoot = resolve(options.repoRoot);
	const defaultOutput = options.adapterName
		? `.agents/cautilus-adapters/${options.adapterName}.yaml`
		: ".agents/cautilus-adapter.yaml";
	const output = resolve(repoRoot, options.output ?? defaultOutput);
	if (existsSync(output) && !options.force) {
		fail(`Adapter already exists at ${output}. Use --force to overwrite.`);
	}
	mkdirSync(dirname(output), { recursive: true });
	const adapter = scaffoldAdapter(repoRoot, options.repoName ?? repoRoot.split(/[\\/]/u).at(-1));
	writeFileSync(output, dumpYamlDocument(adapter), "utf-8");
	process.stdout.write(`${output}\n`);
}

const entryHref = process.argv[1] ? pathToFileURL(resolve(process.argv[1])).href : "";
if (import.meta.url === entryHref) {
	main();
}
